### PR TITLE
feat(#148): Fix Concurrent Problem With Lints Loading

### PR DESCRIPTION
**Pull request by volodya-lombrozo**
In this PR I demonstrate that the current implementation contains one significant concurrency bug.
This bug is related to the usage of the `Sticky` object from the `cactoos` library in a multi-threaded environment without synchronisation.
Therefore, in this PR I:
1. Added a `Lints` abstraction that encapsulates the logic of retrieving all lints (this allows us to test `Lints` independently from `Program`)
2. Added unit tests for `Lints` that demonstrate the problem
3. Fixed the original issue by using the `Synced` object from the `cactoos` library.

Closes: #148.

### CODE REVIEW

**Comment by volodya-lombrozo** [comment]
@yegor256 Could you have a look, please?

**Comment by yegor256** [comment]
@volodya-lombrozo isn't it too many classes? We already have `ProgramLints` and `XslLints`. Now, we will also have `Lints`... Can't we put this functionality into `ProgramLints`? Also, having `Lints` class visible publicly and not implementing any interfaces and coming into `Program` as an argument -- looks a bit wrong to me. WDYT?

**Comment by volodya-lombrozo** [comment]
@yegor256 Sure, I agree. What do you think if we do it in a separate PR? I might add a puzzle for it

**Comment by yegor256** [comment]
@volodya-lombrozo I'm not sure this will work for end-user:

```
public Program(final XML xmir, final Lints lints) {
```

The `Lints` class local, while the ctor is public. Maybe the ctor should be made package-private too?

**Comment by yegor256** [comment]
@volodya-lombrozo I suggest making this PR right, instead of making a puzzle. Here, the problem is not cosmetic, it's design-level issue. Keeping it like this in the code base is dangerous.

**Comment by volodya-lombrozo** [comment]
@yegor256 could you have a look one more time, please?

File: src/main/java/org/eolang/lints/Defect.java [hunk 1]
```diff
@@ -32,12 +32,6 @@
  */
 public interface Defect {
 
-    /**
-     * Rule.
-     * @return Unique name of the rule
-     */
-    String rule();
```
Comment (yegor256) [inline src/main/java/org/eolang/lints/Defect.java:39]: @volodya-lombrozo why? we use this method, in EO, in order to let users know which lint/rule failed

File: src/main/java/org/eolang/lints/LintProgram.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+import org.cactoos.Scalar;
+import org.cactoos.iterable.Joined;
+import org.cactoos.scalar.IoChecked;
+import org.cactoos.scalar.Sticky;
+import org.cactoos.scalar.Synced;
+import org.eolang.lints.comments.AsciiOnly;
+import org.eolang.lints.misc.UnitTestIsNotVerb;
+
+/**
+ * Lints to use.
```
Comment (yegor256) [inline src/main/java/org/eolang/lints/LintProgram.java:42]: @volodya-lombrozo it's a single entity, not a collection, right? Why the name is "lints"?

File: src/main/java/org/eolang/lints/Program.java [hunk 3]
```diff
@@ -43,44 +39,47 @@
...
-        () -> new ProgramLints().value()
-    );
+    private final XML xmir;
 
     /**
-     * The XMIR program to analyze.
+     * The lints to use.
      */
-    private final XML xmir;
+    private final Lint<XML> lints;
```
Comment (yegor256) [inline src/main/java/org/eolang/lints/Program.java:49]: @volodya-lombrozo why `lints` if it's a singular `Lint`?

File: src/main/java/org/eolang/lints/Defect.java (hunk 1 above)
Comment (volodya-lombrozo) [reply src/main/java/org/eolang/lints/Defect.java:39]: @yegor256 As you can see, I was able to remove this method without any failed test. It indicates that we don't need this. Ok, I will return it back with a puzzle.

**Comment by volodya-lombrozo** [comment]
> @volodya-lombrozo looks like you are mixing a few design approached into one. Maybe it's better to fix the problem at hands (the concurrency) issues and do the refactoring (if necessary) in another PR?

I suggested exactly the same [here](https://github.com/objectionary/lints/pull/150#issuecomment-2547489945):

> @yegor256 Sure, I agree. What do you think if we will do it in a separate PR? I might add a puzzle for it

Exactly by the same reasons you mentioned. So, originally it's your decision.

**Comment by yegor256** [comment]
@volodya-lombrozo I just don't see a motivation for the refactoring of such kind. What's wrong with the current design?

**Comment by volodya-lombrozo** [comment]
> @volodya-lombrozo isn't it too many classes? We already have `ProgramLints` and `XslLints`. Now, we will also have `Lints`... Can't we put this functionality into `ProgramLints`? Also, having `Lints` class visible publicly and not implementing any interfaces and coming into `Program` as an argument -- looks a bit wrong to me. WDYT?

@yegor256 It's your words. What I've done:

1. Combined `Lints` (originally created by me in this PR) and `ProgramLints` -> `ProgramLints`
2. Made `ProgramLints` implement `Lint<XML>` for consistency with the design 

That's all.

**Comment by yegor256** [comment]
@volodya-lombrozo let's not do this for now. let's just fix the concurrency issue. design changes we can do in a separate PR. would be great if you submit a ticket for that!

**Comment by volodya-lombrozo** [comment]
Requirement 1
> @volodya-lombrozo isn't it too many classes? We already have `ProgramLints` and `XslLints`. Now, we will also have `Lints`... Can't we put this functionality into `ProgramLints`? Also, having `Lints` class visible publicly and not implementing any interfaces and coming into `Program` as an argument -- looks a bit wrong to me. WDYT?

Requirement 2
> @volodya-lombrozo let's not do this for now. let's just fix the concurrency issue. design changes we can do in a separate PR. would be great if you submit a ticket for that!

@yegor256 How can I do this? There are conflicting requirements. I've already wrote some code to satisfy the **Requirement 1**.   Maybe we can fix the current version of code somehow? Or at least to leave a puzzle to fix current design problems if they are exist.

**Comment by volodya-lombrozo** [comment]
> @volodya-lombrozo it was not a requirement

@yegor256 How that?

> @volodya-lombrozo I suggest making this PR right, instead of making a puzzle. Here, the problem is not cosmetic, it's design-level issue. Keeping it like this in the code base is dangerous.

Actually, you insisted on this. I don't see how I can satisfy conflicting "suggestions" :(

> I believe, you can do this with one line in `Program` file

I couldn't, because I had to test a separate part (originally named `Lints`) on concurrency issues. Please, see, the description of this PR. 

> Also, the tests will be helpful.

I added them from the first commit of this PR. Why are you writing about this?

> Just don't add new classes and don't delete existing one -- this is a requirement

Well, it's probably possible, but ineffective as I've already said:

> I couldn't, because I had to test a separate part (originally named `Lints`) on concurrency issues

____

Can we progress with these changes somehow? I ask you for understanding: you suggested one changes, then ask to return them back. Of course we can do it, but let's add a separate ticket for that, please?

**Comment by yegor256** [comment]
@volodya-lombrozo check this out: 9b87673 Let's extend this block of text with your suggestions, it will help me understand what you are trying to achieve.

**Comment by volodya-lombrozo** [comment]
@yegor256 [Opinionated] The following text seems redundant to me:

```
There are also a few classes that implement `Iterable<Lint>`.
They are supposed to be used only by the `Program` and `Programs`,
and are not supposed to be exposed to the user of the library.
They are responsible for providing a set of lints to be executed,
building them from the information in classpath.
```

These classes are internal components or implementation details, if you prefer. I don't think we should mention them.
Moreover, it seems more appropriate to hide them with `package-private` visibility because of the following statement:

> They are supposed to be used only by the `Program` and `Programs`, and are not supposed to be exposed to the user of the library.

Additionally, we might consider avoiding the use of `Iterable<Lint>` altogether. You may wonder how we could create multiple lints in that case. For this, it would be better to use a separate `X implements Lint<XML>` (not `X implements Iterable<Lint>`). We could call it `CompositeLint`, `LintComposite`, `SeveralLints`, or even `LintProgram` (as I have done in this PR). By doing so, you would simplify the overall solution and make the architecture easier to understand.

The closes analogy is `OptTrain implements Optimization` from `eo` repository that behaves as a single `Optimization`, but actually encapsulates a few of them. 

What do you think?

**Comment by volodya-lombrozo** [comment]
@yegor256 Could you take a look one more time, please?

**Comment by yegor256** [comment]
@rultor merge

**Comment by yegor256** [comment]
@volodya-lombrozo thanks!

**Pull request merged**